File: frontend/src/services/movies.js
import api from "./api";

const normalizeMovie = (movie) => ({
  id: movie.id ?? movie.tmdb_id,
  title: movie.title || "Untitled movie",
  overview: movie.overview || "A great pick for your next watch.",
  poster_url: movie.poster_url || (movie.poster_path ? `https://image.tmdb.org/t/p/w500${movie.poster_path}` : null),
  poster_path: movie.poster_path,
  release_date: movie.release_date || null,
  vote_average: movie.vote_average ?? 0,
  ...movie,
});

const safeRequest = async (request) => {
  try {
    const res = await request();
    return Array.isArray(res?.data) ? res.data.map(normalizeMovie) : [];
  } catch (err) {
    console.error(err);
    return [];
  }
};

// Trending Movies
export const getTrendingMovies = async () => {
  const res = await api.get("/movies/trending");
  return (res.data || []).map(normalizeMovie);
};

// Top Rated Movies
export const getTopRatedMovies = async () => {
  const res = await api.get("/movies/top-rated");
  return (res.data || []).map(normalizeMovie);
};

// Upcoming Movies
export const getUpcomingMovies = async () => {
  const res = await api.get("/movies/upcoming");
  return (res.data || []).map(normalizeMovie);
};

// Search Movies
export const searchMovies = async (query) => {
  const res = await api.get(`/movies/search?query=${query}`);
  return (res.data || []).map(normalizeMovie);
};

// AI Recommendations
export const getRecommendations = async (movieName) => {
  const res = await api.get(
    `/recommend/content?movie_name=${encodeURIComponent(movieName)}`
  );
  return (res.data || []).map(normalizeMovie);
};

// Content-Based Recommendations
export const getContentRecommendations = async (movieName) => {
  const res = await api.get(
    `/recommend/content?movie_name=${encodeURIComponent(movieName)}`
  );
  return (res.data || []).map(normalizeMovie);
};

// Collaborative Filtering
export const getCollaborativeRecommendations = async (movieName) => {
  const res = await api.get(
    `/recommend/collaborative?movie_name=${encodeURIComponent(movieName)}`
  );
  return (res.data || []).map(normalizeMovie);
};

export const getMovieDetails = async (id) => {
  const res = await api.get(`/movies/${id}`);
  return res.data;
};

export const getMovieTrailer = async (id) => {
  const response = await api.get(`/movies/${id}/trailer`);
  return response.data;
};

export const getMovieCast = async (id) => {
  const response = await api.get(`/movies/${id}/cast`);
  return response.data;
};

export const getSimilarMovies = async (id) => {
  const response = await api.get(`/movies/${id}/similar`);
  return response.data;
};


export const saveWatchHistory = async (userId, tmdbId) => {
  const res = await api.post("/watch-history", null, {
    params: {
      user_id: userId,
      tmdb_id: tmdbId,
    },
  });

  return res.data;
};

export const saveSearchHistory = async (userId, query) => {
  const res = await api.post("/search-history/", null, {
    params: {
      user_id: userId,
      query: query,
    },
  });

  return res.data;
};

// Personalized AI Recommendations
export const getPersonalizedRecommendations = async (userId) => {
  const res = await api.get(`/recommended/personalized/${userId}`);
  return (res.data || []).map(normalizeMovie);
};

File: backend/app/routes/movies.py
import datetime
from fastapi import APIRouter, HTTPException,Depends
from sqlalchemy.orm import Session
from app.database.connection import get_db
from app.services.tmdb.movies import (
    get_trending_movies,
    get_top_rated_movies,
    get_upcoming_movies,
    search_movies,
)
from app.services.tmdb.details import tmdb_get
from app.services.ai_recommendation_services import get_personalized_recommendations

router = APIRouter(prefix="/movies", tags=["Movies"])


def format_movies(movies):
    formatted = []

    for movie in movies:
        formatted.append({
            "id": movie.get("id"),
            "title": movie.get("title"),
            "overview": movie.get("overview"),
            "poster_url": (
                f"https://image.tmdb.org/t/p/w500{movie['poster_path']}"
                if movie.get("poster_path")
                else None
            ),
            "backdrop_url": (
                f"https://image.tmdb.org/t/p/original{movie['backdrop_path']}"
                if movie.get("backdrop_path")
                else None
            ),
            "release_date": movie.get("release_date"),
            "vote_average": movie.get("vote_average"),
        })

    return formatted


@router.get("/trending")
async def trending_movies():
    try:
        results = get_trending_movies()
        return format_movies(results)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/top-rated")
async def top_rated_movies():
    try:
        results = get_top_rated_movies()
        return format_movies(results)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/upcoming")
async def upcoming_movies():
    try:
        results = get_upcoming_movies()
        today = datetime.date.today()
        future_movies = []

        for movie in results:
            release_date = movie.get("release_date")
            if not release_date:
                continue

            try:
                release_dt = datetime.date.fromisoformat(release_date)
            except ValueError:
                continue

            if release_dt >= today:
                future_movies.append(movie)

        future_movies.sort(key=lambda movie: movie.get("release_date") or "")
        return format_movies(future_movies)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/search")
async def search(query: str):
    try:
        results = search_movies(query)
        return format_movies(results)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/{movie_id}")
async def get_movie_details(movie_id: int):
    try:
        movie = tmdb_get(f"movie/{movie_id}")

        return {
            "id": movie.get("id"),
            "title": movie.get("title"),
            "overview": movie.get("overview"),
            "poster_url": (
                f"https://image.tmdb.org/t/p/w500{movie['poster_path']}"
                if movie.get("poster_path")
                else None
            ),
            "backdrop_url": (
                f"https://image.tmdb.org/t/p/original{movie['backdrop_path']}"
                if movie.get("backdrop_path")
                else None
            ),
            "release_date": movie.get("release_date"),
            "vote_average": movie.get("vote_average"),
            "runtime": movie.get("runtime"),
            "genres": movie.get("genres", []),
            "tagline": movie.get("tagline"),
            "status": movie.get("status"),
            "homepage": movie.get("homepage"),
        }

    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))  


@router.get("/{movie_id}/trailer")
async def get_movie_trailer(movie_id: int):
    try:
        data = tmdb_get(f"movie/{movie_id}/videos")

        videos = data.get("results", [])

        trailer = next(
            (
                video
                for video in videos
                if video.get("site") == "YouTube"
                and video.get("type") == "Trailer"
            ),
            None,
        )

        if not trailer:
            return {"url": None}

        return {
            "url": f"https://www.youtube.com/watch?v={trailer['key']}"
        }

    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))   


@router.get("/{movie_id}/cast")
async def get_movie_cast(movie_id: int):
    try:
        data = tmdb_get(f"movie/{movie_id}/credits")

        cast = []

        for actor in data.get("cast", [])[:10]:
            cast.append({
                "id": actor.get("id"),
                "name": actor.get("name"),
                "character": actor.get("character"),
                "profile_url": (
                    f"https://image.tmdb.org/t/p/w300{actor['profile_path']}"
                    if actor.get("profile_path")
                    else None
                ),
            })

        return cast

    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))   


@router.get("/{movie_id}/similar")
async def get_similar_movies(movie_id: int):
    try:
        data = tmdb_get(f"movie/{movie_id}/similar")

        movies = []

        for movie in data.get("results", [])[:10]:
            movies.append({
                "id": movie.get("id"),
                "title": movie.get("title"),
                "poster_url": (
                    f"https://image.tmdb.org/t/p/w500{movie['poster_path']}"
                    if movie.get("poster_path")
                    else None
                ),
                "vote_average": movie.get("vote_average"),
                "release_date": movie.get("release_date"),
            })

        return movies

    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))        

@router.get("/recommended/{user_id}")
def recommended_movies(
    user_id: int,
    db: Session = Depends(get_db),
):
    movies = get_personalized_recommendations(user_id, db)

    return [
        {
            "id": movie.tmdb_id,
            "tmdb_id": movie.tmdb_id,
            "title": movie.title,
            "poster_url": (
                                f"https://image.tmdb.org/t/p/w500{movie['poster_path']}"
                                if movie.get("poster_path")
                                else None
                            ),
            "overview": movie.overview,
            "vote_average": movie.vote_average or 0,
            "release_date": movie.release_date,
        }
        for movie in movies
    ]